src/coercions/date.typ

```diff
@@ -0,0 +1,58 @@
+/// An attempt is made to convert string, numeric, or dictionary inputs into datetime objects
+///
+/// #example[```
+/// #let schema = z.date(
+///   pre-transform: z.coerce.date
+/// )
+///
+/// #z.parse(2020, schema) \
+/// #z.parse("2020-03-15", schema) \
+/// #z.parse("2020/03/15", schema) \
+/// #z.parse((year: 2020, month: 3, day: 15), schema) \
+/// ```]
+#let date(self, it) = {
+  if (type(it) == type(datetime.today())) {
+    return it
+  }
+  if (type(it) == int) {
```

Comment:
Really unsure about this one, seems more surprising than useful to me.

Reply:
This was just a file being moved, definitely worth being looked at but I'm not sure if its for this PR (which deals with just the version coercion).

On the topic of datetime parsing, I think dictionaries should be supported, on ISO formats, ISO 8601? I wouldn't support offsets but other than that, that should be doable

Reply:
Yeah, dictionaries sound fine to me.

> ISO 8601?

Yeah, the full year, full month and full day must be specified for a string to be a valid date, i.e. `"07"` or `"30-5"` are rejected, because they are ambiguous, `"1970-01-01"` or `"1970-1-1"` are valid. No further checks like checking for the max day of month, that's up to the impl of `datetime` and more complicated than necessary.